### PR TITLE
Simplify test running in CI and locally

### PR DESCRIPTION
**Pull request by rotu**
Now you should be able to get meaningful results by simply running `jest`. Or `npx majestic` if you’re fancy and want a GUI.

Also, IDE integration with Jest plugins should work better, since now it is unambiguous which configuration applies to which test.

Note that render tests are not currently included since they don't currently run under Jest.

Finally, I consolidated a bunch of testing scripts so you don’t get a bajillion notifications for each push to a PR branch.

### CODE REVIEW

**Comment by rotu** [comment]
@HarelM wanna take a look at this? I got annoyed enough with getting preexisting tests to run that I cleaned things up.

While I’m not in love with Jest, I think this makes the current test landscape a lot more manageable.

File: .github/workflows/test-all.yml [hunk 1]
```diff
@@ -0,0 +1,75 @@
+name: Combined tests
```
Comment (HarelM) [inline .github/workflows/test-all.yml:1]: The idea behind splitting the tests is:
1. Have them run in parallel
2. In case something fails there's the ability to re-run only it and it will be relatively faster.

This breaks this concept and is not where we would like to go.

File: package.json [hunk 2]
```diff
@@ -155,17 +154,17 @@
...
-    "test-style-spec": "jest --roots ./test/integration/style-spec",
-    "test-browser": "jest -c ./jest.config.e2e.ts --roots ./test/integration/browser",
-    "test-render": "ts-node test/integration/render/render.test.ts",
-    "test-query": "jest -c ./jest.config.e2e.ts --roots ./test/integration/query",
-    "test-expression": "jest --roots ./test/integration/expression",
-    "test-unit": "jest --roots ./src  -c ./jest.config.ts ",
-    "test-unit-ci": "jest --roots ./src  -c ./jest.config.ts --ci --coverage --testLocationInResults",
-    "test-watch-roots": "jest --watch  -c ./jest.config.ts --roots ",
+    "test": "run-s lint lint-css lint-docs jest",
+    "test-render": "ts-node test/integration/render/run_render_tests.ts",
```
Comment (HarelM) [inline package.json:158]: I would consider keeping the original order to allow better review, other than that, this looks OK.

File: rollup.config.style-spec.ts [hunk 3]
```diff
@@ -4,30 +4,24 @@ import {RollupOptions} from 'rollup';
...
         json(),
-        resolve({
-            browser: true,
-            preferBuiltins: false,
-            // Some users reference modules within style-spec package directly, instead of the bundle
-            // This means that files within the style-spec package should NOT import files from the parent maplibre-gl-js tree.
-            // This check will cause the build to fail on CI allowing these issues to be caught.
-            jail: 'src/style-spec/',
-        }),
+        resolve({rootDir: styleSpecRootDir}),
```
Comment (HarelM) [inline rollup.config.style-spec.ts:24]: Is this the equivalent of the previous `jail`?

File: src/style-spec/package.json [hunk 4]
```diff
@@ -27,14 +25,13 @@
     "gl-style-validate": "bin/gl-style-validate.js",
     "gl-style-format": "bin/gl-style-format.js"
   },
-  "dependencies": {
+  "sideEffects": false,
+  "devDependencies": {
```
Comment (HarelM) [inline src/style-spec/package.json:29]: Why was this changed? These seems like dependencies, am I missing anything?

File: .github/workflows/test-all.yml (hunk 1 above)
Comment (rotu) [reply .github/workflows/test-all.yml:1]: 1. Jobs are run in parallel
2. You can rerun a single failed job in a workflow.

Look at the results of a workflow run in the github UI - I think you'll be pleasantly surprised at the difference between separated as workflows vs separated as jobs.

File: test/integration/query/query.test.ts [hunk 5]
```diff
@@ -131,10 +131,7 @@ function performQueryOnFixture(fixture)  {
 describe('query tests', () => {
 
     beforeAll(async () => {
-
-        browser = await chromium.launch({
-            headless: false,
-        });
+        browser = await chromium.launch();
```
Comment (HarelM) [inline test/integration/query/query.test.ts:134]: Why not headless?

File: rollup.config.style-spec.ts (hunk 3 above)
Comment (rotu) [reply rollup.config.style-spec.ts:24]: No. The previous jail did not work as intended - it marked the dependencies as external, not caused the build to fail.

I'm pretty sure these were the issues this was intended to fix:
https://github.com/maplibre/maplibre-gl-js/issues/2054

Once resolved, I'd like to enforce this from happening again by enabling the eslint [import/no-relative-packages rule](https://github.com/import-js/eslint-plugin-import/blob/HEAD/docs/rules/no-relative-packages.md)

File: test/integration/query/query.test.ts [hunk 6]
```diff
@@ -150,42 +147,36 @@ describe('query tests', () => {
...
-
-            await page.goto(`file:${path.join(__dirname, 'assets/loadMap.html')}`);
-
-            const currentTestName = testName;
-            const fixture = fixtures[currentTestName];
-
-            const style = fixture.style;
-            const options = style.metadata.test;
+    test.each(Object.keys(fixtures))('%s', async  (testName) => {
+        await page.goto(`file:${path.join(__dirname, 'assets/loadMap.html')}`);
```
Comment (HarelM) [inline test/integration/query/query.test.ts:151]: The console.log helps seeing progress when this is ran locally, does `test.each` does the same?

File: test/build/style-spec.test.ts [hunk 7]
```diff
@@ -1,51 +1,20 @@
...
-
-                    if (
-                        slashOrDot ||
-                        windowsFullPath ||
-                        isBuiltin(id) ||
-                        /node_modules/.test(importer!)
-                    ) {
-                        return null;
-                    }
+    test('files build', async () => {
```
Comment (HarelM) [inline test/build/style-spec.test.ts:4]: Can you clarify the difference between the previous test and this one?

File: .github/workflows/test-all.yml (hunk 1 above)
Comment (HarelM) [reply .github/workflows/test-all.yml:1]: I see, this is relatively new, it didn't exist when we split the workflow.

File: jest.config.ts [hunk 8]
```diff
@@ -1,25 +1,78 @@
...
-    transformIgnorePatterns: [],
-    setupFiles: [
-        'jest-canvas-mock',
-        './test/unit/lib/web_worker_mock.ts'
-    ],
+    // any tests that operate on dist files shouldn't compile them again.
+    transformIgnorePatterns: ['<rootDir>/dist']
+} as Partial<Config>;
+
+const config: Config = {
```
Comment (HarelM) [inline jest.config.ts:14]: While I understand the idea behind moving all the configuration to a single file, these project and their definitions are cumbersome - it feels like who ever will add a test might fail just due to this configuration and the files that were selected to run in a specific configuration as it's not clear which directory is responsible for what and these exclude and include patterns seems random...

File: src/style-spec/package.json (hunk 4 above)
Comment (rotu) [reply src/style-spec/package.json]: Oops! Some of them are; and some of them are handled by the rollup scripts in the parent package.

**Comment by HarelM** [comment]
"non-render tests" is too generic, the integration, browser, unit separation is important to understand which tests failed.
There's also a need to update a lot of the readme in this repo to reflect the changes you made to allow running some of the tests etc, maybe even add what you wrote above using `jest` `majestic` etc.

File: test/integration/query/query.test.ts (hunk 6 above)
Comment (rotu) [reply test/integration/query/query.test.ts:151]: Yes.

File: jest.config.ts (hunk 8 above)
Comment (rotu) [reply jest.config.ts:14]: Yes, it is cumbersome (for now). The "projects" are based entirely on what setup is required for a test to succeed. First step is to get all the config in one place. Second step is to smooth out the differences.

File: test/build/style-spec.test.ts (hunk 7 above)
Comment (rotu) [reply test/build/style-spec.test.ts:4]: This test *just* makes sure that the output files are as expected.

The previous test made sure that imports in bundling the style-spec were declared in the project dependencies (but this didn't work anyway, seeing as the versions of `@mapbox/unitbezier` and `json-stringify-pretty-compact` didn't actually match the versions in the style spec's package.json. And also since the build scripts are all run in the context of the parent package.

We *can* separate out the package further and I think we should. I'll note this in #2054

File: test/integration/query/query.test.ts (hunk 6 above)
Comment (rotu) [reply test/integration/query/query.test.ts]: ouch. I just looked into this some more and I see what you mean. While you can see the individual test results, it's really bad at incremental progress, especially when running tests from multiple files. The default test reporter seems to not report test progress until the suite is done. The log messages come out synchronously when this is the only test file (I think only when the tests are run in-band).

It seems unlikely that this is going to be fixed in Jest:
https://github.com/facebook/jest/issues/6616
https://github.com/facebook/jest/issues/4110

**Comment by HarelM** [comment]
@rotu where are we on this?
I see a requirement to add windows to the render test CI in the following PR: #2106 which might collide with this PR

**Comment by zhangyiatmicrosoft** [comment]
> @rotu where are we on this? I see a requirement to add windows to the render test CI in the following PR: #2106 which might collide with this PR

Clarification --- #2106 is to make test-build pass on windows, not ready to add windows to CI (yet), which needs more stabilization work

**Comment by rotu** [comment]
@HarelM Still in draft and a little stalled out. Don't worry about stepping on my toes here.

**Comment by rotu** [comment]
> "non-render tests" is too generic, the integration, browser, unit separation is important to understand which tests failed. There's also a need to update a lot of the readme in this repo to reflect the changes you made to allow running some of the tests etc, maybe even add what you wrote above using `jest` `majestic` etc.

I think you should be able to run the tests generically, but dig into failures with specificity. I don't actually expect users to care about render vs non-render tests locally, since both should be able to run non-interactively on a local machine.

Is this still an issue?

**Comment by HarelM** [comment]
So the generation of the rest data is done in the rest, right?

File: .github/workflows/test-all.yml [hunk 9]
```diff
@@ -0,0 +1,86 @@
+name: Automated Testing
+
+on:
+  push:
+    branches: [main]
+  pull_request:
+  workflow_dispatch:
+
+env:
+  jest: npx jest --reporters default --reporters github-actions
```
Comment (HarelM) [inline .github/workflows/test-all.yml:10]: I would like the commands to be in the packaje.json file under test-ci-something and have this run only commands using npm run ...

**Comment by HarelM** [comment]
This looks good in general, we can always improve according to feedback.

**Comment by rotu** [comment]
> So the generation of the rest data is done in the test, right?

Yep. The test data is not so much “generated” as pulled together from a bunch of separate folders. I got rid of the extra step so now editing the test cases will show up in the test results without having to run that script!

File: .github/workflows/test-all.yml (hunk 9 above)
Comment (rotu) [reply .github/workflows/test-all.yml]: I updated it to `npm run jest-ci`. Is that sufficient?

**Review by HarelM — APPROVED** [review]
Thanks! If there are issues we can always change it in the future...

**Comment by HarelM** [comment]
@rotu did this PR remove coverage from the CI? I don't see any coverage report in the CI in recent PRs...

**Comment by rotu** [comment]
@HarelM I don't think we had any coverage reports before this PR. Could you link one of the coverage reports you're talking about?

**Comment by HarelM** [comment]
It was part of the CI, reported to the console.
